kuzzleio/documentation-V2
# Use new documentation hierarchy

## What does this PR do?

This PR is a refactor of the test framework written by @Njuelle (great work dude :wink: ) according to the new documentation hierarchy. (See [slack](https://kaliop.slack.com/archives/GBJ9VRA01/p1533223318000487))  
```
|- src
    |- sdk-reference
        |- js
            |- 5
            |- 6
        |- php
            |- 3
            |- 4
```
Each SDK version is maintained in his own directory, this involve a lot of documentation duplication but it will be simpler to maintain.  

This PR seems HUGE because I had remove the old directory structure and create the new one. I also have run prettier on many files by accident because I forgot to stop the auto prettier mode.

You have to use the script `run-snippet-tests.sh` to  execute the snippet tests. This script take a path with the argument `-p`. All snippets within this path will be tested. But this path has to contain only a specific sdk version. (eg: `src/sdk-reference/js/6` will work and `src/sdk-reference/js` won't)

You can see the list of supported sdk and versions in `test/sdk-versions.yml`. This file also  specify where to find the SDK version on Git or S3 (https://github.com/kuzzleio/sdk-go/pull/199):
  - latest: `master` branch
  - unstable: `x-dev` branch (used when SDK version is not released yet)
  - stable: `x-stable` branch

The test framework is executed inside Docker images that provide all the necessary tools to run the tests. There is one image per language.  
The test execution follow this workflow: 
  - Run the image containing all the build and test tools: `run_test.sh`
  - Run the test framework inside the image: `test/snippet-testing`, : `test/lib/testManager.js`
  - Download the SDK version matching the snippets to test (this is inferred with the path of the snippets): `test/lib/sdk/`
  - Inject the snippet in the template glue code: `test/lib/snippet.js`
  - Run language linter: `test/lib/runners/`
  - Run the code and check the standard output: `test/lib/runners/`
  - Generate JSON report: `test/lib/helpers/logger.js`

### How should this be manually tested?

  - Step 1 : Run Kuzzle `bash .travis/start_kuzzle.sh`
  - Step 2 : Run all test for js sdk 6 `bash run_test.sh -n -p src/sdk-reference/js/6`
  - Step 3 : Run only tests of index controller `bash run_test.sh -n -p src/sdk-reference/js/6/index`
  - Step 4 : Run all tests `bash run-all-snippet-tests.sh`

### Other changes

  - Invalidate Cloudfront cache only if travis tests succeed
  - Remove the scaffolding tool
  - Allow `console.log` and unused variable in eslint
  - Add controller name in the `yml` files that describe tests to avoid collision in json report.
  - Replace `[code-example=]` tag by `[snippet=]`
  - Use a single entrypoint for all docker images
